README.md
# Windows Sandbox Tools

![sandbox](images/sandbox.jpg)

[![PSGallery Version](https://img.shields.io/powershellgallery/v/WindowsSandboxTools.png?style=for-the-badge&label=PowerShell%20Gallery)](https://www.powershellgallery.com/packages/WindowsSandboxTools/) [![PSGallery Downloads](https://img.shields.io/powershellgallery/dt/WindowsSandboxTools.png?style=for-the-badge&label=Downloads)](https://www.powershellgallery.com/packages/WindowsSandboxTools/)

This repository is a collection of PowerShell tools and scripts that I use to run and configure the Windows Sandbox feature that was introduced in Windows 10 2004. Many of the commands in this repository were first demonstrated on my [blog](https://jdhitsolutions.com/blog/powershell/7621/doing-more-with-windows-sandbox/). I strongly recommend you read the blog post before trying any of the code. As I mention in the blog post, most of the code here will __reduce the security__ of the Windows Sandbox application. This is a trade-off I am willing to make for the sake of functionality that meets *my* requirements. You have to decide how much of the code you would like to use.

__*All code is offered as-is with no guarantees. Nothing in this repository should be considered production-ready or used in critical environments without your extensive testing and validation.*__

## Installing the Windows Sandbox

You need to have the 2004 version of Windows 10 or later. The Windows Sandbox will run a containerized version of your operating system. I don't know off-hand if it is supported on Windows 10 Home or Education editions. Otherwise, you should be able to run these PowerShell commands to get started:

```powershell
Get-WindowsOptionalFeature -Online -FeatureName Containers-DisposableClientVM
Enable-WindowsOptionalFeature -Online -FeatureName Containers-DisposableClientVM
```

## The WindowsSandBoxTools Module

I have created a PowerShell module called `WindowsSandBoxTools`. In this module are functions designed to make it easier to view a `wsb` configuration, create a new configuration, and export a configuration to a file. The module functions use several PowerShell class definitions.

You can read the Microsoft documentation on creating a Windows Sandbox configuration at [https://docs.microsoft.com/windows/security/threat-protection/windows-sandbox/windows-sandbox-configure-using-wsb-file](https://docs.microsoft.com/windows/security/threat-protection/windows-sandbox/windows-sandbox-configure-using-wsb-file). The module commands are designed to abstract the process of creating the XML configuration files which have a `.wsb` extension.

Install the module from the PowerShell Gallery.

```powershell
Install-Module WindowsSandboxTools
```

The module should work in Windows PowerShell and PowerShell 7 on Windows platforms.

### [Start-WindowsSandbox](docs/Start-WindowsSandbox.md)

The [`Start-WindowsSandbox`](docs/Start-WindowsSandbox.mx) function is my primary tool. It has an alias of `wsb`. You can specify the path to the wsb file.

![Start-WindowsSandbox](images/start-windowssandbox.png)

If you use the `NoSetup` parameter, it will launch the default Windows Sandbox. In either usage, you can specify display dimensions for the sandbox. The `WindowSize` parameter expects an array of width and height, like 1024,768. My default is 1920,1080. You may have to drag the window slightly to force the sandbox to redraw the screen and remove the horizontal scrollbar. Setting the display is tricky, and I don't know if what I am using will work for everyone, so if you don't get the results you expect, please post an issue.

The sandbox will start and then minimize while any configurations you may have are executed.

### [Get-WsbConfiguration](docs/Get-WsbConfiguration.md)

If you have created custom configurations, or intend to use the samples from this module, `Get-WsbConfiguration` will display information about the configuration.

```text
PS C:\> Get-WsbConfiguration d:\wsb\simple.wsb
WARNING: No value detected for LogonCommand. This may be intentional on your part.


   Name: Simple

vGPU                 : Enable
MemoryInMB           : 8192
AudioInput           : Default
VideoInput           : Default
ClipboardRedirection : Default
PrinterRedirection   : Default
Networking           : Default
ProtectedClient      : Default
LogonCommand         :
MappedFolders        : C:\scripts -> C:\scripts [RO:False]
```

The command uses a custom format file to display the configuration. I have also found a way to insert metadata into the wsb file which (so far) doesn't appear to interfere with the Windows Sandbox application.

```text
PS C:\> Get-WsbConfiguration d:\wsb\simple.wsb -MetadataOnly

Author     Name   Description                                       Updated
------     ----   -----------                                       -------
Jeff Hicks Simple a simple configuration with mapping to C:\Scripts 12/10/2021 8:50:03 AM
```

### [New-WsbConfiguration](docs/New-WsbConfiguration.md)

This is how you can create a new Windows Sandbox configuration.

```powershell
$params = @{
 Networking = "Default"
 LogonCommand = "c:\data\demo.cmd"
 MemoryInMB = 2048
 PrinterRedirection = "Disable"
 MappedFolder = (New-WsbMappedFolder -HostFolder d:\data -SandboxFolder c:\data -ReadOnly)
 Name = "MyDemo"
 Description = "A demo WSB configuration"
}
$new = New-WsbConfiguration  @params
```

The `LogonCommand` value is relative to the WindowsSandbox. This code will create a `wsbConfiguration` object.

```text

   Name: MyDemo

vGPU                 : Default
MemoryInMB           : 2048
AudioInput           : Default
VideoInput           : Default
ClipboardRedirection : Default
PrinterRedirection   : Disable
Networking           : Default
ProtectedClient      : Default
LogonCommand         : c:\data\demo.cmd
MappedFolders        : d:\data -> c:\data [RO:True]
```

You could modify this object as necessary.

```powershell
$new.vGPU = "Enable"
$new.Metadata.Updated = Get-Date
```

The last step is to export the configuration to a `wsb` file.

```powershell
$new | Export-WsbConfiguration -Path d:\wsb\demo.wsb
```

Which will create this file:

```xml
<Configuration>
  <Metadata>
    <Name>MyDemo</Name>
    <Author>Jeff</Author>
    <Description>A demo WSB configuration</Description>
    <Updated>07/10/2022 15:40:54</Updated>
  </Metadata>
  <vGPU>Enable</vGPU>
  <MemoryInMB>2048</MemoryInMB>
  <AudioInput>Default</AudioInput>
  <VideoInput>Default</VideoInput>
  <ClipboardRedirection>Default</ClipboardRedirection>
  <PrinterRedirection>Disable</PrinterRedirection>
  <Networking>Default</Networking>
  <ProtectedClient>Default</ProtectedClient>
  <LogonCommand>
    <Command>c:\data\demo.cmd</Command>
  </LogonCommand>
  <MappedFolders>
    <MappedFolder>
      <HostFolder>d:\data</HostFolder>
      <SandboxFolder>c:\data</SandboxFolder>
      <ReadOnly>True</ReadOnly>
    </MappedFolder>
  </MappedFolders>
</Configuration>
```

I can easily launch this configuration.

```powershell
Start-WindowsSandbox -Configuration D:\wsb\demo.wsb
```

### [New-WsbMappedFolder](docs/New-WsbMappedFolder.md)

Use this command to create a mapped folder object.

```powershell
$map = New-WsbMappedFolder -HostFolder c:\work -SandboxFolder c:\work
```

Once created, you can use this in a new configuration.

```powershell
New-WsbConfiguration -Name work -MappedFolder $map -Description "Work sandbox"
```

### [Export-WSBConfiguration](docs/Export-WsbConfiguration.md)

After you have created a new configuration, you will want to save it to a file.

```powershell
New-WsbConfiguration -Name work -LogonCommand "$wsbScripts\basic.cmd" -MemoryInMB (4096*2) -MappedFolder (New-WsbMappedFolder -HostFolder d:\work -SandboxFolder c:\work) -Description "My work WSB configuration" | Export-WsbConfiguration -Path "$wsbConfigPath\work.wsb"
```

The exported file must have a `.wsb` file extension. Save the file to your `$wsbConfigPath` location.

You should get a file like this:

```xml
<Configuration>
  <Metadata>
    <Name>work</Name>
    <Author>Jeff</Author>
    <Description>My work WSB configuration</Description>
    <Updated>07/18/2022 15:43:24</Updated>
  </Metadata>
  <vGPU>Default</vGPU>
  <MemoryInMB>8192</MemoryInMB>
  <AudioInput>Default</AudioInput>
  <VideoInput>Default</VideoInput>
  <ClipboardRedirection>Default</ClipboardRedirection>
  <PrinterRedirection>Default</PrinterRedirection>
  <Networking>Default</Networking>
  <ProtectedClient>Default</ProtectedClient>
  <LogonCommand>
    <Command>C:\scripts\windowssandboxtools\wsbScripts\basic.cmd</Command>
  </LogonCommand>
  <MappedFolders>
    <MappedFolder>
      <HostFolder>c:\work</HostFolder>
      <SandboxFolder>c:\work</SandboxFolder>
      <ReadOnly>False</ReadOnly>
    </MappedFolder>
  </MappedFolders>
</Configuration>
```

## My Configuration Scripts

My default [configuration script](wsbScripts/demo-config.ps1) takes about 4 minutes to complete. I use the [BurntToast](https://github.com/Windos/BurntToast) module to show a Windows Action Center notification when it is complete.

When you import the module, it will define two global variables. `$wsbConfigPath` points to the location of your wsb files. `$wsbScripts` points to the location of your supporting scripts. By default, these variables will point to directories in the module root. You will most like want to update these variables to point to your location. I set this in my PowerShell profile script.

__You need to verify and update path references if you use any of the configurations or scripts in this module.__

## RoadMap

This is a list of items I'd like to address or handle more efficiently:

+ Look for a way to organize script components used for `LogonCommand` settings.
+ Use a default shared folder that can be a bit more generic.
+ Add a better way to provide or integrate toast notifications.
+ Create a graphical interface for creating configurations.

Feel free to post enhancment suggestions in Issues.


changelog.md
# Change Log for Windows SandBox Tools

## v1.1.0

### Changed

- Updated license.
- General code cleanup.
- Updated help documentation.
- Modified `Get-WSBConfiguration` to accept paths from the pipeline.

### Added

- Added missing online help links

## v1.0.0

+ Updated `Start-WindowsSandbox` to not scale resolutions.
+ Fixed bug in `Start-WindowsSandbox` that was testing for a configuration when using the basic configuration. [Issue 4](https://github.com/jdhitsolutions/WindowsSandboxTools/issues/4)
+ Reorganized module layout.
+ Added my configurations and scripts.
+ Moved class definitions to the root module file.
+ Updated help documentation.
+ Updated `New-WsbMappedFolder` to use a switch for `-Readonly`.
+ Fixed bug where Verbose remained on if importing the module with `-Verbose`.
+ Updated `README.md`.
+ Published to the PowerShell Gallery

## December 14, 2020

+ Code cleanup for the sake of clarity.
+ Migrated commands to a module, `WindowsSandboxTools`.
+ Made `Configuration` parameter mandatory in `Start-WindowsSandbox`.
+ Fixed default logo path in `RegisterWatcher`.
+ Added timestamp to verbose messages in `Start-WindowsSandbox`.
+ Modified `Register-Watcher` to remove any left-over event subscriber.
+ Created a global variable `wsbConfigPath` which can be used to reference a centralized location for WSB files.
+ Created a global variable `wsbScripts` which can be used to reference a centralized location for scripts.
+ Added argument completer for the `Configuration` parameter in `Start-WindowsSandbox` and `Get-WsbConfiguration`.
+ Updated `Start-WindowsSandbox` to test if an instance is already running and display a warning if it is.
+ Updated setup scripts and sample configuration files.

## September 10, 2020

+ Changed default screen size in `Start-WindowsSandbox` to 1920x1080.
+ Created commands, `Get-WsbConfiguration`, `New-WsbConfiguration`, `New-WsbMappedFolder`, and `Export-WsbConfiguration` for creating and managing `wsb` files. The enhanced files can include metadata information. These functions are defined in a module file, `wsbFunctions.psm1`.
+ Created `wsbConfiguration.format.ps1xml` to display the `wsbConfiguration` object as a formatted list by default.
+ Fixed `Start-Process` parameter bug in `Start-WindowsSandbox`.
+ Modified `RegisterWatcher.ps1` to let the user define the flag file. The default is `C:\scripts\sandbox.flag`.
+ Modified `RegisterWatcher.ps1` to let the user define settings for the toast notifications. Defaults are mine.
+ Updated `Start-WindowsSandbox` to better handle `-Whatif` support.
+ Updated `README.md`

## August 31, 2020

+ Added code to minimize Windows Sandbox window when running a configuration. (Issue #2)
+ Minor tweaks to the WSB configuration files.
+ Added `-WhatIf` support to `Start-WindowsSandbox` and `RegisterWatcher.ps1`.
+ Added `Set-WindowState` function.
+ Modified `Sandbox-toast.ps1` to make the toast announcement silent.
+ Added Sound to local toast notification.
+ Added code to adjust the window size. This the closest I can get to setting a desktop resolution. (Issue #1)
+ Added presentation configurations and setup scripts.

## August 27, 2020

+ initial commit
+ Added `README.md`
+ Updated scripts to reflect the new path.
+ Added a file system watcher to send a toast notification when my configuration is complete.


functions/public.ps1

Class wsbMetadata {
    [string]$Author = $env:USERNAME
    [string]$Name
    [string]$Description
    [datetime]$Updated

    wsbMetadata([string]$Name) {
        $this.Name = $Name
    }
    wsbMetadata([string]$Name, [string]$Description) {
        $this.Name = $Name
        $this.Description = $Description
    }
}

Class wsbMappedFolder {
    [string]$HostFolder
    [string]$SandboxFolder
    [bool]$ReadOnly = $True

    wsbMappedFolder([string]$HostFolder, [string]$SandboxFolder, [bool]$ReadOnly) {
        $this.HostFolder = $HostFolder
        $this.SandboxFolder = $SandboxFolder
        $this.ReadOnly = $ReadOnly
    }
}

Class wsbConfiguration {
    [ValidateSet("Default", "Enable", "Disable")]
    [string]$vGPU = "Default"
    [string]$MemoryInMB = 4096
    [ValidateSet("Default", "Enable", "Disable")]
    [string]$AudioInput = "Default"
    [ValidateSet("Default", "Enable", "Disable")]
    [string]$VideoInput = "Default"
    [ValidateSet("Default", "Disable")]
    [string]$ClipboardRedirection = "Default"
    [ValidateSet("Default", "Enable", "Disable")]
    [string]$PrinterRedirection = "Default"
    [ValidateSet("Default", "Disable")]
    [string]$Networking = "Default"
    [ValidateSet("Default", "Enable", "Disable")]
    [string]$ProtectedClient = "Default"
    [string]$LogonCommand
    [wsbMappedFolder[]]$MappedFolder
    [wsbMetadata]$Metadata
}

Function New-WsbMappedFolder {
    [cmdletbinding()]
    [OutputType("wsbMappedFolder")]
    Param(
        [Parameter(Position = 0, Mandatory, ValueFromPipelineByPropertyName, HelpMessage = "Specify the path to the local folder you want to map.")]
        [ValidateScript( { Test-Path $_ })]
        [string]$HostFolder,

        [Parameter(Position = 1, Mandatory, ValueFromPipelineByPropertyName, HelpMessage = "Specify the mapped folder for the Windows Sandbox. It must start with C:\")]
        [ValidateNotNullorEmpty()]
        [ValidatePattern('^C:\\')]
        [string]$SandboxFolder,

        [Parameter(ValueFromPipelineByPropertyName, HelpMessage = "Specify if you want the mapping to be Read-Only.")]
        [ValidateSet("True", "False")]
        [string]$ReadOnly = "True"
    )

    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN] Starting $($myinvocation.mycommand)"
    } #begin
    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Processing"

        #turn the strings into booleans
        if ($ReadOnly -eq 'true') {
            $read = $True
        }
        else {
            $read = $False
        }
        [wsbMappedFolder]::New($HostFolder, $SandboxFolder, $Read)
    } #process
    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"
    } #end
}

Function New-WsbConfiguration {
    [cmdletbinding(DefaultParameterSetName = "name")]
    [outputType("wsbConfiguration")]
    Param(
        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Enable", "Disable")]
        [string]$vGPU = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateScript( { $_ -ge 1024 })]
        [string]$MemoryInMB = 4096,

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Enable", "Disable")]
        [string]$AudioInput = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Enable", "Disable")]
        [string]$VideoInput = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Disable")]
        [string]$ClipboardRedirection = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Enable", "Disable")]
        [string]$PrinterRedirection = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Disable")]
        [string]$Networking = "Default",

        [Parameter(ValueFromPipelineByPropertyName)]
        [ValidateSet("Default", "Enable", "Disable")]
        [string]$ProtectedClient = "Default",

        [Parameter(ValueFromPipelineByPropertyName, HelpMessage = "The path and file are relative to the Windows Sandbox")]
        [object]$LogonCommand,

        [Parameter(ValueFromPipelineByPropertyName)]
        [wsbMappedFolder[]]$MappedFolder,

        [Parameter(ParameterSetName = "meta")]
        [wsbMetadata]$Metadata,

        [parameter(Mandatory, ParameterSetName = "name", HelpMessage = "Give the configuration a name")]
        [string]$Name,

        [parameter(ParameterSetName = "name", HelpMessage = "Provide a description")]
        [string]$Description,

        [parameter(ParameterSetName = "name", HelpMessage = "Who is the author?")]
        [string]$Author = $env:USERNAME
    )

    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN] Starting $($myinvocation.mycommand)"
        $new = [wsbConfiguration]::new()

    } #begin
    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Creating a configuration"
        Write-Verbose ($psboundparameters | Out-String)

        if ($logonCommand -is [string]) {
            $cmd = $LogonCommand
        }
        elseif ($logonCommand.Command) {
            $cmd = $logonCommand.Command
        }
        else {
            Write-Warning "No value detected for LogonCommand. This may be intentional on your part."
            $cmd = $Null
        }

        $new.vGPU = $vGPU
        $new.MemoryInMB = $MemoryInMB
        $new.AudioInput = $AudioInput
        $new.VideoInput = $VideoInput
        $new.ClipboardRedirection = $ClipboardRedirection
        $new.PrinterRedirection = $PrinterRedirection
        $new.Networking = $Networking
        $new.ProtectedClient = $ProtectedClient
        $new.LogonCommand = $cmd
        $new.MappedFolder = $MappedFolder

        if (-Not $metadata) {
            $metadata = [wsbMetadata]::new($Name, $Description)
            $metadata.author = $author
            $metadata.updated = Get-Date
        }

        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Setting metadata"
        $new.metadata = $metadata
        $new

    } #process
    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"

    } #end
}
Function Get-WsbConfiguration {
    [cmdletbinding()]
    [outputType("wsbConfiguration")]
    Param(
        [Parameter(Position = 0, Mandatory, HelpMessage = "Specify the path to the .wsb file.")]
        [ValidatePattern('\.wsb$')]
        [ValidateScript({Test-Path $_ })]
        [ArgumentCompleter({if (Test-Path $global:wsbConfigPath) {(Get-Childitem $Global:wsbConfigPath).fullname}})]
        [string]$Path,
        [Parameter(HelpMessage = "Only display metadata information.")]
        [switch]$MetadataOnly
    )

    $cPath = Convert-Path $Path
    Write-Verbose "Getting configuration from $cPath"

    [xml]$wsb = Get-Content -path $cPath

    Write-Verbose "Building property list"
    $properties = [System.Collections.Generic.List[object]]::new()
    $properties.AddRange($wsb.configuration.psadapted.psobject.properties.name)
    if ($properties.Contains("LogonCommand")) {
        [void]($properties.Remove("LogonCommand"))
        $properties.Add(@{Name = "LogonCommand"; Expression = { $_.logonCommand.Command } })
    }
    if ($properties.Contains("MappedFolders")) {
        [void]($properties.Remove("MappedFolders"))
        $properties.Add(@{Name = "MappedFolder"; Expression = { $_.MappedFolders.MappedFolder | New-WsbMappedFolder } })
    }

    Write-Verbose "Get metadata information"
    if ($wsb.configuration.metadata) {
        Write-Verbose "Existing metadata found"
        $meta = [wsbMetadata]::new($wsb.Configuration.metadata.Name, $wsb.configuration.metadata.Description)
        $meta.updated = $wsb.configuration.metadata.updated
        $meta.author = $wsb.configuration.metadata.author
    }
    else {
        Write-Verbose "Defining new metadata"
        $meta = [wsbMetadata]::new($cPath)
        $meta.updated = Get-Date
    }

    if ($MetadataOnly) {
        Write-Verbose "Displaying metadata only"
        $meta
    }
    else {
        Write-Verbose "Sending configuration to New-WsbConfiguration"
        $wsb.configuration | Select-Object -property $properties | New-WsbConfiguration -metadata $meta
        Write-Verbose "Ending $($myinvocation.mycommand)"
    }
}

function Export-WsbConfiguration {
    [cmdletbinding(SupportsShouldProcess)]
    [outputType([System.io.fileInfo])]
    param (
        [Parameter(Mandatory, Position = 0, ValueFromPipeline, HelpMessage = "Specify a wsbConfiguration object.")]
        [ValidateNotNullorEmpty()]
        [wsbConfiguration]$Configuration,
        [Parameter(Mandatory, HelpMessage = "Specify the path and filename. It must end in .wsb")]
        [ValidatePattern("\.wsb$")]
        #verify the parent path exists
        [ValidateScript( { Test-Path $(Split-Path $_ -parent) })]
        [string]$Path,
        [Parameter(HelpMessage = "Don't overwrite an existing file")]
        [switch]$NoClobber
    )
    Begin {
        Write-Verbose "[$((Get-Date).TimeofDay) BEGIN] Starting $($myinvocation.mycommand)"
        <#
        Convert the path to a valid FileSystem path. Normally, I would use Convert-Path
        but it will fail if the file doesn't exist. Instead, I'll split the path, convert
        the root and recreate it.
        #>
        $parent = Convert-Path (Split-Path -path $Path -Parent)
        $leaf = Split-Path -Path $Path -leaf
        $cPath = Join-Path -Path $parent -ChildPath $leaf
    } #begin
    Process {
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Creating an XML document"
        $xml = New-Object System.Xml.XmlDocument
        $config = $xml.CreateElement("element", "Configuration", "")

        #add my custom metadata
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Inserting metadata"
        $meta = $xml.CreateElement("element", "Metadata", "")
        $metaprops = "Name", "Author", "Description", "Updated"
        foreach ($prop in $metaprops) {
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] ..$prop = $($configuration.Metadata.$prop)"
            $item = $xml.CreateElement($prop)
            $item.set_innertext($configuration.Metadata.$prop)
            [void]$meta.AppendChild($item)
        }
        [void]$config.AppendChild($meta)

        #add primary properties
        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Adding primary data"
        $mainprops = "vGPU", "MemoryInMB", "AudioInput", "VideoInput", "ClipboardRedirection", "PrinterRedirection",
        "Networking", "ProtectedClient"
        foreach ($prop in $mainprops) {
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] ..$prop = $($configuration.$prop)"
            $item = $xml.CreateElement($prop)
            $item.set_innertext($configuration.$prop)
            [void]$config.AppendChild($item)
        }
        if ($Configuration.LogonCommand) {
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Adding a logon command"
            $logon = $xml.CreateElement("element", "LogonCommand", "")
            $cmd = $xml.CreateElement("Command")
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] ..$($configuration.LogonCommand)"
            $cmd.set_innerText($configuration.LogonCommand)
            [void]$logon.AppendChild($cmd)
            [void]$config.AppendChild($logon)
        }
        if ($Configuration.MappedFolder) {
            Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Adding mapped folders"
            $mapped = $xml.CreateElement("element", "MappedFolders", "")
            foreach ($f in $Configuration.MappedFolder) {
                Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] $($f.HostFolder) -> $($f.SandboxFolder) [RO:$($f.ReadOnly)]"
                $mf = $xml.CreateElement("element", "MappedFolder", "")
                "HostFolder", "SandboxFolder", "ReadOnly" |
                    ForEach-Object {
                        $item = $xml.CreateElement($_)
                        $item.set_innertext($f.$_)
                        [void]$mf.AppendChild($item)
                    }
                [void]$mapped.appendChild($mf)
            }
            [void]$config.AppendChild($mapped)
        }

        [void]$xml.AppendChild($config)

        Write-Verbose "[$((Get-Date).TimeofDay) PROCESS] Exporting configuration to $cPath"
        if ($PSCmdlet.ShouldProcess($cpath)) {
            if ((Test-Path -path $cPath) -AND $NoClobber) {
                Write-Warning "A file exists at $cPath and NoClobber was specified."
            }
            else {
                $xml.Save($cPath)
            }
        } #whatif

    } #process
    End {
        Write-Verbose "[$((Get-Date).TimeofDay) END    ] Ending $($myinvocation.mycommand)"

    } #end
}

Function Start-WindowsSandbox {
    [cmdletbinding(SupportsShouldProcess, DefaultParameterSetName = "config")]
    [alias("wsb")]
    [OutputType("none")]

    Param(
        [Parameter(Position = 0, Mandatory,ParameterSetName = "config", HelpMessage = "Specify the path to a wsb file.")]
        [ValidateScript({ Test-Path $_ })]
        [ArgumentCompleter({if (Test-Path $global:wsbConfigPath) {(Get-Childitem $Global:wsbConfigPath).fullname}})]
        [string]$Configuration,

        [Parameter(ParameterSetName = "normal", HelpMessage = "Start with no customizations.")]
        [switch]$NoSetup,
        
        [Parameter(HelpMessage = "Specify desktop resolutions as an array like 1280,720. The default is 1920,1080.")]
        [int[]]$WindowSize = @(1920, 1080)
    )

    Write-Verbose "[$((Get-Date).TimeOfDay)] Starting $($myinvocation.mycommand)"

    #test if Windows Sandbox is already running
    $test = Get-Process -ErrorAction SilentlyContinue -Name WindowsSandbox*
    if ($test) {
        Write-Warning "An instance of Windows Sandbox is already running."
        #bail out
        return
    }
    if ($NoSetup) {
        Write-Verbose "[$((Get-Date).TimeOfDay)] Launching default WindowsSandbox.exe"
        if ($PSCmdlet.shouldProcess("default configuration", "Launch Windows Sandbox")) {
            Start-Process -filepath $env:windir\system32\WindowsSandbox.exe
        }
    }
    else {
        Write-Verbose "[$((Get-Date).TimeOfDay)] Launching WindowsSandbox using configuration file $Configuration"

        #TODO - Parameterize this behavior
        #create a file watcher if calling a configuration that uses it
        if ($Configuration -match "WinSandBx|presentation") {
            Write-Verbose "[$((Get-Date).TimeOfDay)] Registering a temporary file system watcher"
            <#
            Some configurations will create "flag" file in a shared folder to indicate
             the setup and configuration is complete.

            RegisterWatcher is a private function in this module
            #>
            # TODO : pass arguments to RegisterWatcher from this function ?
            [void](Register-Watcher)
        }

        if ($PSCmdlet.shouldProcess($Configuration, "Launch Windows Sandbox")) {
            Start-Process -filepath $Configuration
        }
    }

    #WindowsSandbox.exe launches a child process, WindowsSandboxClient.
    #That is the process that needs to be minimized
    if ($pscmdlet.shouldProcess("WindowsSandboxClient", "Waiting for process")) {

        Write-Verbose "[$((Get-Date).TimeOfDay)] Waiting for child process WindowsSandboxClient"
        do {
            Start-Sleep -Seconds 1
        } Until (Get-Process -Name WindowsSandboxClient -ErrorAction SilentlyContinue)

        #give the process a chance to complete
        Start-Sleep -Seconds 5
    }

    Write-Verbose "[$((Get-Date).TimeOfDay)] Setting Window size to $($WindowSize -join 'x')"
    if ($pscmdlet.shouldProcess("WindowsSandboxClient", "Modifying Window size and state")) {
        $clientProc = Get-Process -Name WindowsSandboxClient
        #values to pass to the function are a percentage of the desired size
        [int]$width = $WindowSize[0] * .6667
        [int]$Height = $WindowSize[1] * .6667
        Write-Verbose "[$((Get-Date).TimeOfDay)] Modifying handle $($clientproc.mainwindowhandle) with a width of $($WindowSize[0]) and height of $($WindowSize[1])"

        Set-WindowSize -handle $clientproc.MainWindowHandle -width $width -height $height

        #Configure the Windows Sandbox to run minimized (Issue #2)
        Write-Verbose "[$((Get-Date).TimeOfDay)] Minimizing child process WindowsSandboxClient"
        $clientProc | Set-WindowState -State minimize
    }

    $wsb = Get-WsbConfiguration -Path $Configuration
    if ($wsb.LogonCommand) {
        Write-Verbose "[$((Get-Date).TimeOfDay)] Running logon command $($wsb.LogonCommand) in the Windows Sandbox."
    }

    if ($wsb.MappedFolder) {
        Write-Verbose "[$((Get-Date).TimeOfDay)] Mapping these folders"
        $wsb.MappedFolder | Out-String | Write-Verbose
    }
    Write-Verbose "[$((Get-Date).TimeOfDay)] Ending $($myinvocation.mycommand)."

    Write-Host "[$((Get-Date).TimeOfDay)] Windows Sandbox has been launched. You may need to wait for any configurations to complete." -foregroundColor green
}
